### PR TITLE
Driver module parameter verification scripts for mlx5_ib

### PR DESCRIPTION
**Pull request by manvanthar**
Driver module parameter verification scripts for mlx5_ib

Signed-off-by: Manvanthara Puttashankar <manvanth@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by manvanthar** [comment]
avocado run driver_parameter.py -m driver_parameter.py.data/driver_parameter_mlx5_ib.yaml 
JOB ID     : a8b5b7ff7c5f1b5f6e55447ede17874578e0d19a
JOB LOG    : /root/tests/results/job-2019-11-11T08.20-a8b5b7f/job.log
 (1/2) driver_parameter.py:Moduleparameter.test;dc_cnak_qp_depth-a5cf: PASS (81.76 s)
 (2/2) driver_parameter.py:Moduleparameter.test;dc_cnak_qp_depth_100-ee8b: PASS (81.70 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 163.55 s
JOB HTML   : /root/tests/results/job-2019-11-11T08.20-a8b5b7f/results.html

**Comment by manvanthar** [comment]
[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/3831611/job.log)

File: io/driver/driver_parameter.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        checking whether the given module is built_in module or not
+        """
+        path = "/lib/modules/%s/modules.builtin" % self.uname
+        for each in genio.read_all_lines(path):
+            out = process.getoutput(each.split('/')[-1])
+            if module == out.split('.'[0]):
+                return True
+            return False
+
+    def ping_check(self, options):
```
Comment (narasimhan-v) [inline io/driver/driver_parameter.py]: Please use https://github.com/avocado-framework/avocado/blob/master/avocado/utils/configure_network.py#L108

File: io/driver/driver_parameter.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            linux_modules.load_module(mod1)
+            time.sleep(self.load_unload_sleep_time)
+        sub_mod = linux_modules.get_submodules(mod1)
+        if sub_mod:
+            for mod in sub_mod.split(' '):
+                linux_modules.unload_module(mod)
+                if linux_modules.module_is_loaded(mod) is True:
+                    self.error_modules.append(mod)
+                    break
+        cmd = "rmmod %s" % mod1
```
Comment (narasimhan-v) [inline io/driver/driver_parameter.py]: Please use https://github.com/avocado-framework/avocado/blob/master/avocado/utils/linux_modules.py#L162

File: io/driver/driver_parameter.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        if process.system(cmd, shell=True, verbose=True,
+                          ignore_status=True) != 0:
+            self.fail("Unloading Module %s failed" % mod1)
+        time.sleep(self.load_unload_sleep_time)
+        location = process.system_output('/sbin/modinfo -n %s' %
+                                         self.module).decode('utf-8')
+        cmd1 = "insmod %s %s=%s" % \
```
Comment (narasimhan-v) [inline io/driver/driver_parameter.py]: Please use https://github.com/avocado-framework/avocado/blob/master/avocado/utils/linux_modules.py#L46

File: io/driver/driver_parameter.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                self.fail("Sysfs check failed ")
+        if not self.ping_check("-c 1000 -f"):
+            self.fail("ping test failed")
+
+    def test(self):
+        """
+        Test Begins here
+        """
+        if self.built_in_module(self.module) is True:
+            self.cancel("Module %s is Built-in Skipping " %
```
Comment (narasimhan-v) [inline io/driver/driver_parameter.py]: This can be done in `setUp()`

File: io/driver/driver_parameter.py (hunk 1 above)
Comment (manvanthar) [reply io/driver/driver_parameter.py]: ok can be done!!

File: io/driver/driver_parameter.py (hunk 2 above)
Comment (manvanthar) [reply io/driver/driver_parameter.py]: There is an catch.

unload_module uses the modinfo -r method of removing....    If this is being used.. We cannot use insmod for sending in parameters.  

The actual way of working is
#rmmod driver_name
#insmod driver_name

or

#modprobe -r driver_name
#modprobe driver_name

btw even modinfo driver_name <parameter>=<value> wont work...   It has to be rmmod and then insmod.

File: io/driver/driver_parameter.py (hunk 3 above)
Comment (manvanthar) [reply io/driver/driver_parameter.py]: explaination same as above^^ ...insmod has to be followed by rmmod.    Enhancing exsisting function might break the other scripts.

@narasimhan-v

File: io/driver/driver_parameter.py (hunk 4 above)
Comment (manvanthar) [reply io/driver/driver_parameter.py]: ok can be done!!!

File: io/driver/driver_parameter.py (hunk 2 above)
Comment (narasimhan-v) [reply io/driver/driver_parameter.py]: What is the reason `modprobe driver_name parameter=value` won't work @manvanthar ? Because, that is the ideal way

Comment (manvanthar) [reply io/driver/driver_parameter.py]: @narasimhan-v ... Here is the output... Modprobe is mainly used to pick the default parameters.

# modprobe -r mlx5_ib
# modprobe mlx5_ib dc_cnak_qp_depth=1500
# cat /sys/module/mlx5_ib/parameters/dc_cnak_qp_depth 
8192

Comment (narasimhan-v) [reply io/driver/driver_parameter.py]:
```
# cat /sys/module/kvm_hv/parameters/one_vm_per_core 
N
# rmmod kvm_hv
# modprobe kvm_hv one_vm_per_core=Y
# cat /sys/module/kvm_hv/parameters/one_vm_per_core 
Y
#
```

Comment (manvanthar) [reply io/driver/driver_parameter.py]: hmm current libraries uses modprobe -r.    Should that be changed to rmmod?

Comment (narasimhan-v) [reply io/driver/driver_parameter.py]: `modprobe -r` or `rmmod` is fine. I was just demonstrating `modprobe module param=value` works fine.

Comment (manvanthar) [reply io/driver/driver_parameter.py]: Changes included in latest code!!

File: io/driver/driver_parameter.py (hunk 4 above)
Comment (manvanthar) [reply io/driver/driver_parameter.py]: Changes included in latest code!!

File: io/driver/driver_parameter.py (hunk 3 above)
Comment (manvanthar) [reply io/driver/driver_parameter.py]: Changes included in latest code!!

File: io/driver/driver_parameter.py (hunk 1 above)
Comment (manvanthar) [reply io/driver/driver_parameter.py]: Changes included in latest code!!

**Comment by manvanthar** [comment]
@narasimhan-v   run after the latest code changes

avocado run driver_parameter.py -m driver_parameter.py.data/driver_parameter_mlx5_ib.yaml 
Fetching assets from driver_parameter.py:Moduleparameter.test
JOB ID     : 1ae4fa0db03599ec615751fab2fdef07fc6a4baf
JOB LOG    : /root/tests/results/job-2019-11-20T07.07-1ae4fa0/job.log
 (1/2) driver_parameter.py:Moduleparameter.test;dc_cnak_qp_depth-1e36: PASS (102.97 s)
 (2/2) driver_parameter.py:Moduleparameter.test;dc_cnak_qp_depth_100-d611: PASS (100.52 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 203.58 s
JOB HTML   : /root/tests/results/job-2019-11-20T07.07-1ae4fa0/results.html

**Comment by manvanthar** [comment]
[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/3869153/job.log)

File: io/driver/driver_parameter.py.data/driver_parameter_mlx5_ib.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+netmask:
+peer_ip:
+sysfs_check_required: 0
+module_param: !mux
+    dc_cnak_qp_depth:
+        module_param_name: "dc_cnak_qp_depth"
+        module_param_value: "1"
+    dc_cnak_qp_depth_100:
+        module_param_name: "dc_cnak_qp_depth"
+        module_param_value: "100"
```
Comment (narasimhan-v) [inline io/driver/driver_parameter.py.data/driver_parameter_mlx5_ib.yaml]: How about this:
```
module_param_name: "dc_cnak_qp_depth"
    value: !mux
        1:
            module_param_value: "1"
        100:
            module_param_value: "1"
```

File: io/driver/driver_parameter.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            self.fail("Unloading Module %s failed" % mod1)
+        time.sleep(self.load_unload_sleep_time)
+        cmd = "%s %s=%s" % (mod1, self.param_name, self.param_value)
+        if linux_modules.load_module(cmd) is False:
+            self.fail("Param %s = Value %s Failed for Module %s" %
+                      (self.param_name, self.param_value, mod1))
+        if self.sysfs_chk:
+            if self.sysfs_value_check() is False:
+                self.fail("Sysfs check failed ")
+        if not HostInfo.ping_check(self, self.ifaces, self.peer, '1000',
```
Comment (narasimhan-v) [inline io/driver/driver_parameter.py]: @manvanthar we should set IP Addr here as well, using `configure_network.set_ip`

File: io/driver/driver_parameter.py.data/driver_parameter_mlx5_ib.yaml (hunk 5 above)
Comment (manvanthar) [reply io/driver/driver_parameter.py.data/driver_parameter_mlx5_ib.yaml]: Done,latest commit has this change

**Comment by manvanthar** [comment]
[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/3890278/job.log)

File: io/driver/driver_parameter.py (hunk 6 above)
Comment (narasimhan-v) [reply io/driver/driver_parameter.py]: @manvanthar is this resolved ?

Comment (manvanthar) [reply io/driver/driver_parameter.py]: @narasimhan-v   my comments are missing :(

configure_network.set_ip is not needed before ping_check.   As it is covered in Setup.  setting ifcfg is an one time activity which is done in Setup.  

After which , module load and unload just does ifup to pick the IP address.

Comment (narasimhan-v) [reply io/driver/driver_parameter.py]: Ah sure, thanks @manvanthar

**Comment by narasimhan-v** [comment]
@manvanthar please squash into a single commit, and we are ready to merge.

**Comment by manvanthar** [comment]
Fixed in PR #1450 1450

**Pull request closed without merging**